flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandler.java

```diff
@@ -0,0 +1,297 @@
+
+		try {
+			StateInitializationContext initializationContext = new StateInitializationContextImpl(
+				context.isRestored(), // information whether we restore or start for the first time
+				operatorStateBackend, // access to operator state backend
+				keyedStateStore, // access to keyed state backend
+				keyedStateInputs, // access to keyed state stream
+				operatorStateInputs); // access to operator state stream
+
+			initializeOperatorAction.accept(initializationContext);
+		} finally {
+			closeFromRegistry(operatorStateInputs, closeableRegistry);
+			closeFromRegistry(keyedStateInputs, closeableRegistry);
+		}
+	}
+
+	private static void closeFromRegistry(Closeable closeable, CloseableRegistry registry) {
+		if (registry.unregisterCloseable(closeable)) {
+			IOUtils.closeQuietly(closeable);
+		}
+	}
+
+	public void dispose() throws Exception {
+		try (Closer closer = Closer.create()) {
+			if (closeableRegistry.unregisterCloseable(operatorStateBackend)) {
+				closer.register(operatorStateBackend);
+			}
+			if (closeableRegistry.unregisterCloseable(keyedStateBackend)) {
+				closer.register(keyedStateBackend);
+			}
+			if (operatorStateBackend != null) {
+				closer.register(() -> operatorStateBackend.dispose());
+			}
+			if (keyedStateBackend != null) {
+				closer.register(() -> keyedStateBackend.dispose());
+			}
+		}
+	}
+
+	public OperatorSnapshotFutures snapshotState(
+			ThrowingConsumer<StateSnapshotContext, Exception> snapshotStateAction,
+			InternalTimeServiceManager<?> timeServiceManager,
+			String operatorName,
+			long checkpointId,
+			long timestamp,
+			CheckpointOptions checkpointOptions,
+			CheckpointStreamFactory factory) throws CheckpointException {
+		KeyGroupRange keyGroupRange = null != keyedStateBackend ?
+			keyedStateBackend.getKeyGroupRange() : KeyGroupRange.EMPTY_KEY_GROUP_RANGE;
+
+		OperatorSnapshotFutures snapshotInProgress = new OperatorSnapshotFutures();
+
+		StateSnapshotContextSynchronousImpl snapshotContext = new StateSnapshotContextSynchronousImpl(
+			checkpointId,
+			timestamp,
+			factory,
+			keyGroupRange,
+			closeableRegistry);
+
+		snapshotState(
+			snapshotStateAction,
+			timeServiceManager,
+			operatorName,
+			checkpointId,
+			timestamp,
+			checkpointOptions,
+			factory,
+			snapshotInProgress,
+			snapshotContext);
+
+		return snapshotInProgress;
+	}
+
+	@VisibleForTesting
+	void snapshotState(
+			ThrowingConsumer<StateSnapshotContext, Exception> snapshotStateAction,
+			InternalTimeServiceManager<?> timeServiceManager,
+			String operatorName,
+			long checkpointId,
+			long timestamp,
+			CheckpointOptions checkpointOptions,
+			CheckpointStreamFactory factory,
+			OperatorSnapshotFutures snapshotInProgress,
+			StateSnapshotContextSynchronousImpl snapshotContext) throws CheckpointException {
+		try {
+			timeServiceManager.snapshotState(keyedStateBackend, snapshotContext, operatorName);
+			snapshotStateAction.accept(snapshotContext);
```

Comment:
Operator state snapshot is taken using `Consumer snapshotStateAction`, 
but for `timeServiceManager` we pass it and call directly.

Should we make it consistent? 

We could remove `opName` parameter and then use `BiConsumerWithException`.
`opName` can be turned into a field or exception can be wrapped on a higher level.

Reply:
Those kind of interfaces and lambda invocations are hampering the code readability and brake IDE support (find usages/implementations). By using `Consumer` I was cutting some corners - avoiding defining some extra dedicated interface, so replacing `timeServiceManager` with `BiConsumerWithException` would be a step back.

If you want I could go the other way, and replace `Consumer` with some proper interface.

Reply:
In terms of readability, it's a tradeoff: I don't know what is going on inside, but I do know how is it used in this method just by looking at its type.
And yes, passing "functions" allows to avoid unnecessary interfaces and coupling.

With or without them, I think consistency is more important.

Reply:
I introduced `StreamOperatorStateHandler.CheckpointedStreamOperator` interface to make it consistent. Can you take a look one more time @rkhachatryan ?